public/script.js

```diff
@@ -8487,99 +8513,143 @@ export function callPopup(text, type, inputValue = '', { okButton, rows, wide, w
+            !(messageId === 0 && !chat_metadata?.tainted &&
+                (message?.['swipes']?.length ?? 1) == 1
+            )
+        )
+    )
+    //The message is swipeable.
+    { return true; }
+    //The message is not swipeable.
+    else { return false; }
+}
 
-    const currentMessage = chatElement.children().filter(`[mesid="${mesId}"]`);
-    const swipeId = chat[mesId].swipe_id;
-    const swipeCounterText = formatSwipeCounter((swipeId + 1), chat[mesId]['swipes'].length);
-    const swipeRight = currentMessage.find('.swipe_right');
-    const swipeLeft = currentMessage.find('.swipe_left');
-    const swipeCounter = currentMessage.find('.swipes-counter');
+/**
+ * Refreshes all swipe buttons and updates their swipe counters.
+ * This has been optimized for bulk updates by minimizing DOM queries.
+ */
+export function refreshSwipeButtons() {
+    //Never show swipe buttons on an empty chat.
+    if (chat?.length === 0) return false;
 
-    if (swipeId !== undefined && (chat[mesId].swipes.length > 1 || swipeId > 0)) {
-        swipeLeft.css('display', 'flex');
-    }
-    //only show right when generate is off, or when next right swipe would not make a generate happen
-    if (is_send_press === false || chat[mesId].swipes.length >= swipeId) {
-        swipeRight.css('display', 'flex').css('opacity', '0.3');
-        swipeCounter.css('opacity', '0.3');
-    }
-    if ((chat[mesId].swipes.length - swipeId) === 1) {
-        //chevron was moved out of hardcode in HTML to class toggle dependent on last_mes or not
-        //necessary for 'swipe_right' div in past messages to have no chevron if 'show swipes for all messages' is turned on
-        swipeRight.css('opacity', '0.7');
-        swipeCounter.css('opacity', '0.7');
-    }
+    //If swipes are disabled or hidden, hide all swipe buttons.
+    if (!isSwipingAllowed()) {
+        $('body').toggleClass('hideAllSwipeButtons', true);
+        return;
+    //Don't hide all swipe buttons.
+    } else {
+        //CSS will hide all messages.
+        $('body').toggleClass('hideAllSwipeButtons', false);
+    }
+    //Non-messages can appear in chat. '.mes' is required.
+    const messageElements = chatElement.children('.mes[mesid]');
+
+    // const lastDisplayedMesId = Number(messageElements.last().attr('mesid'));
+    const firstDisplayedMesId = Number(messageElements.first().attr('mesid'));
+
+    //Group each message.
+    messageElements.each((index, div) => {
+        // const messageId = Number($(div).attr('mesid')); //Slower.
+        //This assumes the messages are in order and their Id's are accurate.
+        const messageId = firstDisplayedMesId + index;
+
+        const message = chat[messageId];
+        if (isMessageSwipeable(messageId, message)) {
+            //If a right swipe would trigger a generation or loop to the first swipe.
+            if ((message?.['swipes']?.length ?? 1) - 1 <= (message?.['swipe_id'] ?? 0 )) {
```

Comment:
I'd prefer direct property access instead of indexing for all the new code
```suggestion
            if ((message?.swipes?.length ?? 1) - 1 <= (message?.swipe_id ?? 0 )) {
```

Reply:
Fixed: https://github.com/SillyTavern/SillyTavern/pull/4712/commits/1867c24c26f6951a4af9c35daae4c9175c3957bc